src/Microsoft.ML.Dnn/DnnCatalog.cs

```diff
@@ -0,0 +1,133 @@
+        /// This parameter is used to deal with models that have unknown shape but the internal operators in the model require data to have batch dimension as well.</param>
+        /// <param name="dnnFramework"></param>
+        /// <remarks>
+        /// The support for retraining is under preview.
+        /// </remarks>
+        public static DnnEstimator RetrainDnnModel(
+            this ModelOperationsCatalog catalog,
+            string[] outputColumnNames,
+            string[] inputColumnNames,
+            string labelColumnName,
+            string tensorFlowLabel,
+            string optimizationOperation,
+            string modelPath,
+            int epoch = 10,
+            int batchSize = 20,
+            string lossOperation = null,
+            string metricOperation = null,
+            string learningRateOperation = null,
+            float learningRate = 0.01f,
+            bool addBatchDimensionInput = false,
+            DnnFramework dnnFramework = DnnFramework.Tensorflow)
+        {
+            var options = new Options()
+            {
+                ModelLocation = modelPath,
+                InputColumns = inputColumnNames,
+                OutputColumns = outputColumnNames,
+                LabelColumn = labelColumnName,
+                TensorFlowLabel = tensorFlowLabel,
+                OptimizationOperation = optimizationOperation,
+                LossOperation = lossOperation,
+                MetricOperation = metricOperation,
+                Epoch = epoch,
+                LearningRateOperation = learningRateOperation,
+                LearningRate = learningRate,
+                BatchSize = batchSize,
+                AddBatchDimensionInputs = addBatchDimensionInput,
+                ReTrain = true
+            };
+
+            var env = CatalogUtils.GetEnvironment(catalog);
+            return new DnnEstimator(env, options, DnnUtils.LoadDnnModel(env, modelPath, true));
+        }
+
+        /// <summary>
+        /// Performs image classification using transfer learning.
+        /// </summary>
+        /// <param name="catalog"></param>
+        /// <param name="featuresColumnName"></param>
+        /// <param name="labelColumnName"></param>
+        /// <param name="outputGraphPath"></param>
+        /// <param name="scoreColumnName"></param>
+        /// <param name="predictedLabelColumnName"></param>
+        /// <param name="checkpointName"></param>
+        /// <param name="arch"></param>
+        /// <param name="dnnFramework"></param>
+        /// <param name="epoch"></param>
+        /// <param name="batchSize"></param>
+        /// <param name="learningRate"></param>
+        /// <param name="addBatchDimensionInput"></param>
+        /// <remarks>
+        /// The support for image classification is under preview.
+        /// </remarks>
+        public static DnnEstimator ImageClassification(
```

Comment:
`ImageClassification `should allow the user to provide their own DNN model (TF model until now) in their parameters, meaning that as an optional parameter the user should be able to provide the `Options `object including the 'ModelLocation` (either on a local drive path or an HTTP path pointing to the source model) and the additional TF model configuration (labelColumn, epoch, etc.) for the chose TF model.

Reply:
No, the design of this API is to not allow the users to add their own models. We have a set if models and they get to pick. There will be a different API that will let you do transfer learning for any model.

Reply:
I understood that in the discussion the other day, that this API would allow additional models.
Why a different API if the 'external' model can also be about ImageClassification?
If it is something very different I understand, but being something pretty similar (another model for ImageClassification) why do we want a different API? We could have optional parameters, too.

Reply:
ImageClassification is specific to ImageClassification scnearios. I want to control the models so that I can control how I edit the graph. If I allow the user to feed in whatever graph they want then I will need to also provide functionality for the user to do graph editing. The plan is we will have such a generic API before GA.

---
In reply to: [310331079](https://github.com/dotnet/machinelearning/pull/4057#discussion_r310331079) [](ancestors = 310331079)